### PR TITLE
File plugin

### PR DESCRIPTION
**Pull request by ksahilCF**
@sawantpritam Pr raise please review

### CODE REVIEW

**Comment by ksahilCF** [comment]
@tivv @rmstar @itsankit-google  please review.

File: e2e-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            <dependency>
+                <groupId>io.cdap.tests.e2e</groupId>
+                <artifactId>cdap-e2e-framework</artifactId>
+                <version>0.0.1-SNAPSHOT</version>
+                <scope>test</scope>
+            </dependency>
+    </dependencies>
+
+    <properties>
+        <maven.compiler.source>17</maven.compiler.source>
```
Comment (tivv) [inline e2e-test/pom.xml]: Why?

File: e2e-test/src/test/java/io/cdap/plugin/utils/GCConstants.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * constants.
+ */
+public class GCConstants {
+  public static final String ERROR_MSG_COLOR = "errorMessageColor";
+  public static final String ERROR_MSG_DATASET = "errorMessageDataset";
+  public static final String ERROR_MSG_TEMP_BUCKET = "errorMessageTemporaryBucket";
+  public static final String ERROR_MSG_REF = "errorMessageReference";
+  public static final String CDF_ATHENA_BQMTABLE = "cdf-athena-bq-bq-multitable";
+  public static final String A = "a";
```
Comment (tivv) [inline e2e-test/src/test/java/io/cdap/plugin/utils/GCConstants.java]: What is this constant about?

File: e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import java.io.IOException;
+import java.util.Properties;
+
+import static io.cdap.plugin.utils.GCConstants.ERROR_MSG_COLOR;
+import static io.cdap.plugin.utils.GCConstants.ERROR_MSG_MANDATORY;
+
+/**
+ * CdapUtils contains the helper functions.
+ */
+public class CdapUtils {
```
Comment (tivv) [inline e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java]: This is too generic name.

File: e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  private static Properties pluginProperties = new Properties();
+  private static Properties errorProperties = new Properties();
+  private static final Logger logger = Logger.getLogger(CdapUtils.class);
+
+  static {
+
+    try {
+      pluginProperties.load(new FileInputStream("src/test/resources/pluginParameters.properties"));
```
Comment (tivv) [inline e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java]: Don't load from source directory. Load from classloader

File: e2e-test/pom.xml (hunk 1 above)
Comment (sawantpritam) [reply e2e-test/pom.xml]: Changed

File: e2e-test/src/test/java/io/cdap/plugin/utils/GCConstants.java (hunk 2 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/utils/GCConstants.java]: Removed unused constants.

File: e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java (hunk 3 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java]: Renamed to `E2ETestUtils`

File: e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java (hunk 4 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/utils/CdapUtils.java]: Changed.

File: e2e-test/src/test/java/io/cdap/plugin/file/locators/package-info.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Package contains the runner for the GCP features.
+ */
+package io.cdap.plugin.file.locators;
+
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/locators/package-info.java]: nit: remove extra newlines

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+import static io.cdap.plugin.utils.E2ETestConstants.ERROR_MSG_ERROR_FOUND_VALIDATION;
+import static io.cdap.plugin.utils.E2ETestConstants.ERROR_MSG_FILE_INVALID_OUTPUTFIELD;
+import static io.cdap.plugin.utils.E2ETestConstants.ERROR_MSG_VALIDATION;
+
+/**
+ * File Connector related Step design.
+ **/
+
+public class FileConnector implements CdfHelper {
+  List<String> propertiesOutputSchema = new ArrayList<String>();
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: new ArrayList<>();

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+    if (property.equalsIgnoreCase("referenceName")) {
+      CdfFileActions.enterFileBucket(E2ETestUtils.pluginProp("fileCsvFilePath"));
+      CdfFileActions.selectFormat(E2ETestUtils.pluginProp("fileCSVFileFormat"));
+    } else if (property.equalsIgnoreCase("path")) {
+      CdfFileActions.enterReferenceName();
+      CdfFileActions.selectFormat(E2ETestUtils.pluginProp("fileCSVFileFormat"));
+    } else if (property.equalsIgnoreCase("format")) {
+      CdfFileActions.enterReferenceName();
+      CdfFileActions.enterFileBucket(E2ETestUtils.pluginProp("fileCsvFilePath"));
+    }
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:94]: fail test if property is not one of the above strings

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+  }
+
+  @Then("Validate mandatory property error for {string}")
+  public void validateMandatoryPropertyErrorFor(String property) {
+    CdfStudioActions.clickValidateButton();
+    SeleniumHelper.waitElementIsVisible(CdfStudioLocators.validateButton);
+    E2ETestUtils.validateMandatoryPropertyError(property);
+  }
+
+  @Then("Connect Source as {string} and sink as {string} to establish connection")
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: nit: Source -> source

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+    CdfStudioActions.connectSourceAndSink(source, sink);
+  }
+
+  @Then("Enter the File connector Properties with file path {string} and format {string}")
+  public void enterTheFileConnectorPropertiesWithFilePathAndFormat(String filePath, String format) throws IOException,
+    InterruptedException {
+    CdfFileActions.enterReferenceName();
+    CdfFileActions.enterFileBucket(E2ETestUtils.pluginProp(filePath));
+    CdfFileActions.selectFormat(E2ETestUtils.pluginProp(format));
+    CdfFileActions.enterSampleSize(E2ETestUtils.pluginProp("fileSampleSize"));
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:115]: why is this hardcoded, not taken as param?

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+    CdfFileActions.skipHeader();
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:116]: do we have tests where we don't skip the header?

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 6 above)
Comment (rmstar) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: also, why is schema a list? shouldn't it be a map of field name -> field type?

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+    Optional<String> result = GcpClient
+      .getSoleQueryResult("SELECT distinct " + E2ETestUtils.pluginProp(outputField) + " as bucket FROM `"
+                            + (E2ETestUtils.pluginProp("projectId")) + "."
+                            + (E2ETestUtils.pluginProp("dataset")) + "."
+                            + E2ETestUtils.pluginProp(targetTable) + "` ");
+    String pathFromBQTable = StringUtils.EMPTY;
+    if (result.isPresent()) {
+      pathFromBQTable = result.get();
+    }
+    BeforeActions.scenario.write("GCC bucket path in BQ Table :" + pathFromBQTable);
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: GCS

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+                            + (E2ETestUtils.pluginProp("dataset")) + ".INFORMATION_SCHEMA.COLUMNS` " +
+                            "WHERE table_name = '" + E2ETestUtils.pluginProp(targetTable)
+                            + "' and column_name = '" + E2ETestUtils.pluginProp(field) + "' ");
+    String dataTypeInTargetTable = StringUtils.EMPTY;
+    if (result.isPresent()) {
+      dataTypeInTargetTable = result.get();
+    }
+    BeforeActions.scenario.write("Data type in target BQ Table :" + dataTypeInTargetTable);
+    Assert.assertEquals(E2ETestUtils.pluginProp(dataType),
+                        dataTypeInTargetTable.replace("64", StringUtils.EMPTY).toLowerCase());
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:387]: why are you doing this?

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+  @Then("Validate successMessage is displayed")
+  public void validateSuccessMessageIsDisplayed() {
+    CdfLogActions.validateSucceeded();
+  }
+
+  @Then("Get Count of no of records transferred to BigQuery in {string}")
+  public void getCountOfNoOfRecordsTransferredToBigQueryIn(String tableName) throws IOException, InterruptedException {
+    int countRecords = GcpClient.countBqQuery(E2ETestUtils.pluginProp(tableName));
+    BeforeActions.scenario.write("**********No of Records Transferred******************:" + countRecords);
+    Assert.assertTrue(countRecords > 0);
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: number of records transferred should be equal to the number of input records

**Comment by rmstar** [comment]
Are we also planning to test File sink plugin?

File: e2e-test/src/test/java/io/cdap/plugin/file/locators/CdfFileLocators.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  public static WebElement fileProperties;
+
+  @FindBy(how = How.XPATH, using = "//*[@data-cy='switch-skipHeader']")
+  public static WebElement skipHeader;
+
+  @FindBy(how = How.XPATH, using = "//*[contains(text(),'Get Schema')]")
+  public static WebElement getSchemaButton;
+
+  @FindBy(how = How.XPATH, using = "//*[@data-cy=\"plugin-File-batchsource\"]")
+  public static WebElement fileBucket;
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/locators/CdfFileLocators.java]: 'bucket' is associated with object storage. Is this referring to a file path?

Comment (itsankit-google) [reply e2e-test/src/test/java/io/cdap/plugin/file/locators/CdfFileLocators.java]: looks like it is referring to file batch source plugin.

Comment (itsankit-google) [reply e2e-test/src/test/java/io/cdap/plugin/file/locators/CdfFileLocators.java]: fileBucket -> fileBatchSource

File: e2e-test/src/test/java/io/cdap/plugin/file/locators/package-info.java (hunk 5 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/locators/package-info.java]: removed extra newlines

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 6 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: Changed.

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 7 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:94]: Added else block to fail test if property is not one of the above strings

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 8 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: Changed

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 9 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:115]: Changed to take it as param

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 10 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:116]: Added test in which we don't skip the header.

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 11 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: changed

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 12 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java:387]: In BQ INFORMATION_SCHEMA table data_type is "FLOAT64" and on CDAP UI it is getting displayed as only "float".

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java (hunk 13 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/FileConnector.java]: Changed assertion

File: e2e-test/src/test/java/io/cdap/plugin/file/locators/CdfFileLocators.java (hunk 14 above)
Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/file/locators/CdfFileLocators.java]: Renamed.

File: e2e-test/pom.xml [hunk 15]
```diff
@@ -1,4 +1,19 @@
 <?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright © 2015-2019 Cask Data, Inc.
```
Comment (rmstar) [inline e2e-test/pom.xml:3]: 2022

File: e2e-test/src/test/java/io/cdap/plugin/file/actions/package-info.java [hunk 16]
```diff
@@ -1,4 +1,20 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/actions/package-info.java:2]: 2022

File: e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/package-info.java [hunk 17]
```diff
@@ -1,4 +1,20 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/stepsdesign/package-info.java:2]: 2022

File: e2e-test/src/test/java/io/cdap/plugin/file/tests/runner/package-info.java [hunk 18]
```diff
@@ -1,4 +1,20 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
```
Comment (rmstar) [inline e2e-test/src/test/java/io/cdap/plugin/file/tests/runner/package-info.java:2]: 2022

File: e2e-test/src/test/features/FileConnectorSchema.feature [hunk 19]
```diff
@@ -1,5 +1,5 @@
 Feature: Validate File plugin output schema for different formats
-  @File
+  @File1
```
Comment (rmstar) [inline e2e-test/src/test/features/FileConnectorSchema.feature:2]: why is this changed to File1?

File: e2e-test/src/test/features/FileConnectorSchema.feature [hunk 20]
```diff
@@ -1,5 +1,5 @@
...
   Scenario Outline:FILE Source output schema validation
```
Comment (rmstar) [inline e2e-test/src/test/features/FileConnectorSchema.feature:3]: we should have separate tests (and directories) for source and sink plugins

**Comment by rmstar** [comment]
please fix merge conflicts

**Comment by ksahilCF** [comment]
Old Pull request. Hence, closing this.
